pos_accounting/src/pool/block_undo.rs

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{btree_map::Entry, BTreeMap};
+
+use crate::PoSAccountingUndo;
+
+use common::{chain::Transaction, primitives::Id};
+use serialization::{Decode, Encode};
+use thiserror::Error;
+
+#[derive(Error, Debug, PartialEq, Eq, Clone)]
+pub enum BlockUndoError {
+    #[error("Attempted to insert a transaction in undo that already exists: `{0}`")]
+    UndoAlreadyExists(Id<Transaction>),
+}
+
+#[derive(Default, Debug, Clone, Eq, PartialEq, Encode, Decode)]
+pub struct TxUndo(Vec<PoSAccountingUndo>);
+
+impl TxUndo {
+    pub fn new(undos: Vec<PoSAccountingUndo>) -> Self {
+        Self(undos)
+    }
+
+    pub fn inner(&self) -> &[PoSAccountingUndo] {
+        &self.0
+    }
+
+    pub fn into_inner(self) -> Vec<PoSAccountingUndo> {
+        self.0
+    }
+}
+
+#[derive(Clone, Debug, Default, Eq, PartialEq, Encode, Decode)]
+pub struct BlockUndo {
```

Comment:
Maybe we should give this a different name? `AccountingBlockUndo`, while the other one `UtxosBlockUndo?`

Reply:
I used them as pos_accounting::BlockUndo and utxo::BlockUndo. But I don't mind to rename

Reply:
In principle it's nice, but the problem is that typically people using IDEs will just import the whole thing when they need it. That's why I'm not really super-excited about being able to distinguish between different types through scopes.